File: python_cpp/uPIMulator_frontend/src/abi/isa/cc/imm_shift_nz_cc.py
from typing import Set

from abi.isa.cc._base_cc import BaseCC
from abi.isa.instruction.condition import Condition


class ImmShiftNZCC(BaseCC):
    def __init__(self, condition: Condition):
        super().__init__(
            ImmShiftNZCC.conditions(), condition,
        )

    @staticmethod
    def conditions() -> Set[Condition]:
        return {
            Condition.Z,
            Condition.NZ,
            Condition.XZ,
            Condition.XNZ,
            Condition.E,
            Condition.O,
            Condition.PL,
            Condition.MI,
            Condition.SZ,
            Condition.SNZ,
            Condition.SPL,
            Condition.SMI,
            Condition.SE,
            Condition.SO,
            Condition.TRUE,
        }

File: golang_vm/uPIMulator/src/device/linker/kernel/instruction/cc/imm_shift_nz_cc.go
package cc

import (
	"errors"
)

type ImmShiftNzCc struct {
	condition Condition
}

func (this *ImmShiftNzCc) Init(condition Condition) {
	conditions := map[Condition]bool{
		Z:    true,
		NZ:   true,
		XZ:   true,
		XNZ:  true,
		E:    true,
		O:    true,
		PL:   true,
		MI:   true,
		SZ:   true,
		SNZ:  true,
		SPL:  true,
		SMI:  true,
		SE:   true,
		SO:   true,
		TRUE: true,
	}

	if _, found := conditions[condition]; !found {
		err := errors.New("condition is not allowed")
		panic(err)
	}

	this.condition = condition
}

func (this *ImmShiftNzCc) Condition() Condition {
	return this.condition
}
